### ISSUE TITLE
Rule: `option_map_or_else_none_and_some`

### ISSUE THREAD

**Issue by tvolk131** [rust-lang/rust-clippy#13303]
### What it does

Very similar to the [`option_map_or_none`](https://rust-lang.github.io/rust-clippy/master/index.html#/option_map_or_none) rule.

When calling `map_or_else()` on an `Option<T>` and the following two conditions are met:

* The `default` parameter is a lambda that always returns `None`
* The `f` parameter is a lambda that always returns `Some`

This is effectively a more verbose way of doing what `Option::map()` does.

### Advantage

Code is shorter, more expressive, and more concise.

### Drawbacks

Given that this rule is only applicable within such narrow constraints, I don't think there is any downside.

### Example

```rust
Some(1).map_or_else(|| None, |num| Some(num + 1))
```

Could both be written as:

```rust
Some(1).map(|num| num + 1)
```

**Comment by nyurik** [comment]
while similar to `option_map_or_none`, this lint would need to modify lambda code, which I would think is slightly trickier:
* `|v| { let i = foo(); Some(i + v) }`
* `|v| if v > 10 { Some(v+1) } else { Some(v-1) }`
* `|v| match v { 1 => Some(1), _ => Some(v+1)  }`
* a combination of the above

In other words, there would need to be code evaluating every code path inside the lambda to check that every one either returns `Some(...)` or `!` (i.e. panics)

**Comment by tvolk131** [comment]
> this lint would need to modify lambda code

Good point! I was thinking solely of lambdas that immediately return `None` or `Some`, but the examples you listed would be nice to tackle too. I imagine the simple case could be done first, and the cases you mentioned could be added later.

**Comment by ada4a** [comment]
This can already be achieved by applying 3 lints one after another:
Starting from `.map_or_else(|| None, |num| Some(num + 1))`:
1. [`unnecessary_lazy_evaluations`](https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations) should turn that into `.map_or(None, |num| Some(num + 1))` -- it currently doesn't, but there's a feature request over at https://github.com/rust-lang/rust-clippy/issues/15829
2. `option_map_or_none` will turn that into `.and_then(|num| Some(num + 1))`
3. [`bind_instead_of_map`](https://rust-lang.github.io/rust-clippy/master/index.html#bind_instead_of_map) will turn that into `.map(|num| num + 1)`

When the feature request is implemented, one will be able to run `cargo clippy --fix`, which would apply fixes one after the other, resulting in the desired output.

**Comment by tvolk131** [comment]
Very nice! That seems like a much more elegant fix.
